### PR TITLE
wip: group join refactor - xstate / dependency control

### PR DESCRIPTION
**Pull request by technoplato**
I promise I will clean up this PR
clean-it-up-mcjuggernuggets

this pr is going to contain a whole bunch of different ideas

it will be broken down into smaller prs after initially sharing with team and determining what we like/don't like. 

GitHub conversation: https://github.com/ephemeraHQ/converse-app/discussions/929

### CODE REVIEW

**Comment by technoplato** [comment]
* **#823** <a href="https://app.graphite.dev/github/pr/ephemeraHQ/converse-app/823?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @technoplato and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: i18n/translations/en.ts [hunk 1]
```diff
@@ -260,13 +260,23 @@ const en = {
   // Group Invites
   group_admin_approval:
     "A group admin may need to approve your membership prior to joining.",
+  "A group admin may need to approve your membership prior to joining.":
```
Comment (alexrisch) [inline i18n/translations/en.ts:263]: This feels like a weird approach, why use full sentences rather than keys?

File: openai/openrouter/openai/o1-preview
Comment (alexrisch) [inline openai/openrouter/openai/o1-preview:1]: This can be deleted

File: queries/useGroupInviteQuery.ts [hunk 2]
```diff
@@ -8,7 +8,11 @@ export const useGroupInviteQuery = (groupInviteId: string) => {
   const account = useCurrentAccount() as string;
 
   return useQuery<GroupInvite | undefined>({
+    // I need to learn more about how query keys work to determine
```
Comment (alexrisch) [inline queries/useGroupInviteQuery.ts:11]: Query keys are just a way to select a certain query status
Technically it can be anything, but it's normally best to combine the dependencies of the function

First we want to have a dependency on the group invite id, so if you look at a different group invite id, it will have a different state

Since the QueryClient is app wide, this means we should also set query keys by account since if you look at the group invite id as one account and then switch to another, it should be a completely different query

File: queries/useGroupInviteQuery.ts [hunk 3]
```diff
@@ -8,7 +8,11 @@ export const useGroupInviteQuery = (groupInviteId: string) => {
...
+    // the significance of this query key with an account
     queryKey: groupInviteQueryKey(account, groupInviteId),
+    // does this need to be in useQuery for any
```
Comment (alexrisch) [inline queries/useGroupInviteQuery.ts:14]: Nope doesn't need to be

Think of useQuery and really all of the query client usage more as a network state wrapper.
The queryFn is just the actual function within the wrapper

I think the majority of XState usage would not actually interact with much of the state of react query other than maybe some prefetches

XState should be our necessary consistent state
React Query should be the helpful things to load later on

If we zoom out a little from this instance and instead look at initial app load
Xstate should handle:
1. Loading the client/checking if that is correct/usable
2. Loading our necessary information (Conversations, Groups, Consent)
3. Optionally it can tell our queries to prefetch some data (like fetching profiles of the first 20 items in our conversation list)

React Query:
React Query handles the actual decision of fetching from the network when needed

If XState was also in charge of fetching all profiles it can become pretty messy and slow

But if React Query handles it, from a DX experience it's a lot cleaner to say:
useProfileData and rely on the caching/persistence offered by the library

Now if we want to look at this use case, we _can_ use React Query just for initial loading of what the group invite network data is, but it's really not providing much other than syntactic sugar on top of useState/useEffect

So I think what makes sense is to remove useQuery entirely from this case and manage the one time load in XState

File: features/GroupInvites/GroupLink.tsx [hunk 4]
```diff
@@ -1,23 +1,22 @@
+import Button from "@components/Button/Button";
```
Comment (alexrisch) [inline features/GroupInvites/GroupLink.tsx:1]: We can just remove this whole screen I believe

File: features/GroupInvites/GroupInvite.screen.tsx [hunk 5]
```diff
@@ -1,81 +1,88 @@
...
   );
+
   const { allowGroup } = useGroupConsent(newGroup?.topic || "");
   const handlingNewGroup = useRef(false);
 
   const handleNewGroup = useCallback(
     async (group: GroupWithCodecsType) => {
-      if (group.client.address === account && !handlingNewGroup.current) {
+      const whatDoesThisConditionIndicate =
+        group.client.address === account && !handlingNewGroup.current;
```
Comment (alexrisch) [inline features/GroupInvites/GroupInvite.screen.tsx:74]: This is checking if the new group is the correct account of what's expected from the user
There's also a check to make sure it's not doing multiple groups
I think both are a bit hacky and can probably be removed

Comment (technoplato) [reply features/GroupInvites/GroupInvite.screen.tsx:74]: Thanks for answering all my questions.

Two follow ups here: 

1) what does `group.client` mean? 

2) When could `client.address` NOT equal the current account?

File: i18n/translations/en.ts (hunk 1 above)
Comment (technoplato) [reply i18n/translations/en.ts:263]: @thierryskoda and I discussed this approach and while at first glance it may seem strange, you should look at it from the perspective of the screen you're building / looking for a screen from a design/app run.

If I try and create a short key for the string, I have to search and find it in this file, and then jump to where that string is used in the code. That's more steps and those steps add up quick since string search is easiest way to find where in the app something is happening. 

cc @thierryskoda for thoughts

File: openai/openrouter/openai/o1-preview
Comment (technoplato) [reply openai/openrouter/openai/o1-preview:1]: Thanks, not sure how that got here at all. No sweat on reviewing this anymore since it's still in draft by the way! Don't want to waste your time w stuff like this.

File: queries/useGroupInviteQuery.ts (hunk 2 above)
Comment (technoplato) [reply queries/useGroupInviteQuery.ts:11]: That totally makes sense - thanks for the explanation. Went through `react-query` docs and they were super helpful, as well as the source. It's pretty clean!

File: queries/useGroupInviteQuery.ts (hunk 3 above)
Comment (technoplato) [reply queries/useGroupInviteQuery.ts:14]: @alexrisch Let me get a little bit more of the code done and we can review it and continue iterating for what should live where from there. It's a little too tricky without my end ready for review first. 

TLDR: 

`xstate` will DEFINITELY not reinvent the wheel that `react-query` has built, so querying/caching/optimistic updating logic will live in `react-query` and `xstate` will react to, observe, and trigger things in `react-query` land. 

Initially, I am going to keep `zustand` as a layer to share state as the transition to `xstate` completely as a state _management_ layer will be more lengthy and there will always be times where we may not want it just to share state in place A and place B. This is a conversation we will continue to revisit to make our workflows as ergonomic as possible. 

> So I think what makes sense is to remove useQuery entirely from this case and manage the one time load in XState

Agree here, although I'm also a huge fan of consistency. SHOULD our entire client -> network communication be done through `queryClient`? My initial gut says yes and that's how I'll build this, but we can always revisit. 

Having consistency to me is more valuable than saving a few keystrokes, and it will be a valuable low risk exercise to understand how to interact with network via `queryClient` in `xstate` within the Controlled Dependencies layer. 

Lots of words, code will help! Stay tuned 📺

File: i18n/translations/en.ts (hunk 1 above)
Comment (thierryskoda) [reply i18n/translations/en.ts:263]: Agree with Michael that this way worked best for me. Especially for finding screen.

File: features/GroupInvites/GroupInvite.screen.tsx (hunk 5 above)
Comment (alexrisch) [reply features/GroupInvites/GroupInvite.screen.tsx:97]: There is a client for each connected account, but I don't think that's really a case in this code, so I think it should be removed

File: features/Inspection/useEventTargetShim.ts
Comment (technoplato) [inline features/Inspection/useEventTargetShim.ts:1]: Required for remote inspector for state machines (stately sky) in expo. Not used yet but keeping it here so it's ready.

File: i18n/i18n.ts [hunk 6]
```diff
@@ -17,6 +17,48 @@ i18n.translations = {
...
+ * // Using English sentences as keys:
+ * i18n.t("Hello, how are you?") // Works as-is
+ * i18n.t("menu::Edit::Cut") // Nested structure
+ *
+ * @sideEffects
+ * - Changes how nested keys are accessed in translations
+ * - Existing translation key strings may need updating
+ * -- Regex to find foo.bar.strings: translate\(.*\..*"\)
+ */
+// export const TranslationNestedSeparator = "::" as const;
```
Comment (technoplato) [inline i18n/i18n.ts:57]: saving this for later

File: jest.config.ts
Comment (technoplato) [inline jest.config.ts:1]: moved config from `package.json#jest` into `jest.config.ts` and added settings to ensure `env` variables were available at test time

File: jest.setup.ts [hunk 7]
```diff
@@ -46,10 +47,10 @@ jest.mock("react-native-fs", () => {
     touch: jest.fn(),
     MainBundlePath: jest.fn(),
     CachesDirectoryPath: jest.fn(),
-    DocumentDirectoryPath: jest.fn(),
```
Comment (technoplato) [inline jest.setup.ts:49]: `string#join` was complaining that we were passing it a function, not a string

File: package.json [hunk 8]
```diff
@@ -61,6 +61,7 @@
     "@sentry/react-native": "^5.31.1",
     "@shopify/flash-list": "1.6.4",
     "@stardazed/streams-polyfill": "^2.4.0",
+    "@statelyai/inspect": "^0.4.0",
```
Comment (technoplato) [inline package.json:64]: not used yet but will be

File: queries/groupConsentMutationUtils.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+export type GroupConsentAction = "allow" | "deny";
+
+export type GroupConsentMutationProps = {
+  account: string;
+  topic: string;
+  groupId: string;
+  action: GroupConsentAction;
+};
+
+export const createGroupConsentMutationObserver = (
```
Comment (technoplato) [inline queries/groupConsentMutationUtils.ts:22]: first stab at using `react-query` imperatively

File: queries/useGroupsQuery.ts [hunk 10]
```diff
@@ -1,18 +1,20 @@
...
 
 import { groupsQueryKey } from "./QueryKeys";
 import { entify, EntityObject } from "./entify";
 import { queryClient } from "./queryClient";
-import {
-  ConverseXmtpClientType,
-  GroupWithCodecsType,
-} from "../utils/xmtpRN/client";
 
-type GroupMembersSelectData = EntityObject<GroupWithCodecsType, string>;
```
Comment (technoplato) [inline queries/useGroupsQuery.ts:13]: string is default 2nd generic, can be removed

File: queries/useGroupsQuery.ts [hunk 11]
```diff
@@ -1,18 +1,20 @@
...
+type GroupMembersSelectData = EntityObject<GroupWithCodecsType>;
 
-const groupsQueryFn = async (account: string) => {
+export const groupsQueryFn = async (account: string): Promise<GroupsEntity> => {
```
Comment (technoplato) [inline queries/useGroupsQuery.ts:17]: prefer explicit type

File: screens/Navigation/SplitScreenNavigation/SplitRightStackNavigation.tsx [hunk 12]
```diff
@@ -109,9 +103,8 @@ export default function SplitRightStackNavigation({
           {WebviewPreviewNav()}
           {ProfileNav()}
           {GroupNav()}
-          {GroupLinkNav()}
```
Comment (technoplato) [inline screens/Navigation/SplitScreenNavigation/SplitRightStackNavigation.tsx:112]: Group link nav was not needed, `GroupInvite` replaced by `JoinGroup` to be more explicit

File: tsconfig.json [hunk 13]
```diff
@@ -1,6 +1,7 @@
 {
   "extends": "expo/tsconfig.base",
   "compilerOptions": {
```
Comment (technoplato) [inline tsconfig.json:3]: required for dynamic imports in `clients`. Will be used to avoid needlessly mocking unused transitive dependencies

File: utils/api.test.ts [hunk 14]
```diff
@@ -28,7 +28,8 @@ jest.mock("./keychain", () => {
   };
 });
 
-describe.skip("GroupInviteLink", () => {
+// todo(lustig): incorporate these tests into JoinGroupClient#live
+describe("GroupInviteLink", () => {
```
Comment (technoplato) [inline utils/api.test.ts:32]: fixed by new `jest.config.ts` see comment there for more details

File: utils/api.ts [hunk 15]
```diff
@@ -17,7 +18,7 @@ import type { ProfileSocials } from "../data/store/profilesStore";
 import type { Frens } from "../data/store/recommendationsStore";
 import { getXmtpApiHeaders } from "../utils/xmtpRN/api";
 
-const api = axios.create({
+export const api = axios.create({
```
Comment (technoplato) [inline utils/api.ts:21]: used to create `JoinGroupClient.live`

**Comment by technoplato** [comment]
closing draft as this was implemented via stacks!
